## Índice

1.   [Identificadores, constantes y variables](#id1)
2.   [Tipos de datos simples. Enteros, reales, booleanos y caracteres](#id2)
3.   [Operadores logicos y de comparacion](#id3)
4.   [Comparaciones encadenadas](#id4)
5.   [Desafio 1: Clasificar números](#des1)
6.   [Desafio 2: Calcular promedio](#des2)
7.   [Desafio 3: Convertidor de temperaturas](#des3)
8.   [Desafio 4: Verificar múltiplos de varios números](#des4)
9.   [Ejemplos](#ejemplos)

<a id="id1"> </a>
# Identificadores, constantes y variables

En Python, un identificador es un nombre que se utiliza para identificar una variable, función, clase, módulo u otro objeto. Los identificadores son sensibles a mayúsculas y minúsculas y deben seguir ciertas reglas de nomenclatura. Para nombrar los identificadores en Python, se utilizan letras, números y guiones bajos. Sin embargo, el primer carácter no puede ser un número y no se permiten espacios ni caracteres especiales.

Para garantizar la creación correcta de identificadores en Python, es importante seguir las PEP 8, que son las guías de estilo de Python. Estas proporcionan recomendaciones sobre cómo nombrar variables, funciones y otros objetos de manera consistente y legible para otros programadores. Puedes encontrar las [PEP 8 en el sitio web oficial de Python](https://peps.python.org/pep-0008/). Siguiendo estas pautas, puedes mejorar la claridad y la legibilidad de tu código, lo que facilita su mantenimiento y colaboración con otros desarrolladores.

Por ejemplo, los siguientes son identificadores válidos en Python:

In [ ]:
numero
total_venta
ListaDeCompras
edad

Mientras que los siguientes no son identificadores válidos:

In [ ]:
2d_array  # el primer carácter no puede ser un número
mi identificador  # no se permiten espacios
total$  # no se permiten caracteres especiales

*Nota 2: Cabe destacar, que por mas que no es inválido, utilizar los caracteres "l" (ele minúscula), "I" (letra i latina mayuscula), "O" (letra o mayúscula) y "0" (número cero), puede causar confusión según el caso en donde se utilicen.*

También se tiene en cuenta que, como la mayoría de lenguajes, Python tiene una lista de palabras reservadas, que no pueden utilizarse como identificadores:

| Palabra reservadas |
|---------------|
| False         |
| True          |
| def           |
| while         |
| if            |
| else          |

Para ver la lista completa utiliza el siguiente codigo en Python:

In [ ]:
import keyword

print(keyword.kwlist)

Las constantes son valores que no cambian durante la ejecución de un programa. En Python, se utilizan variables para almacenar constantes. Las constantes pueden ser números, cadenas de texto, booleanos o cualquier otro valor inmutable.

Por ejemplo, los siguientes son constantes en Python:

In [ ]:
PI = 3.14159
NOMBRE_EMPRESA = "Acme Corporation"
ES_POSITIVO = True

Las variables en Python son objetos que se utilizan para almacenar datos. Las variables se crean cuando se les asigna un valor y pueden cambiar durante la ejecución de un programa.

Por ejemplo, el siguiente código crea una variable llamada edad y le asigna el valor 25:

In [ ]:
edad = 25

En Python, las variables no necesitan ser declaradas antes de usarlas. Python automáticamente asigna un tipo de datos a las variables basándose en el valor que se les asigna. Por ejemplo, si se asigna un número entero a una variable, Python automáticamente la considerará como una variable de tipo int. Si se asigna una cadena de texto a una variable, Python la considerará como una variable de tipo str.

Pero... ¿cómo podemos determinar el tipo de una variable? Si veo "un número" entero, puedo identificar que el tipo de variable es un int o si veo un texto escrito entre comillas (" "), puedo determinar que es una variable str.

In [ ]:
edad = 12
nombre = "Anthony"
altura = 1.75

print ("la variable", edad, "es de tipo:", type(edad))
print ("la variable", nombre, "es de tipo:", type(nombre))
print ("la variable", altura, "es de tipo:", type(altura))


Es importante tener en cuenta que las variables en Python son sensibles a mayúsculas y minúsculas. Esto significa que las variables edad y Edad son dos variables diferentes en Python.

Además, las variables en Python pueden ser eliminadas utilizando la palabra clave del (siempre y cuando esten definidas). Por ejemplo, el siguiente código elimina la variable edad y nombre:

In [ ]:
del edad, nombre
print(altura)


Cuidado! Que si intento eliminar la variable edad, sin crearla de nuevo, dará error, ya que la misma no existe. Este codigo resuelve este problema:


In [ ]:
if "edad" in globals():
    del edad
    print("puedo borrar edad")
else:
    print("no puedo borrar")

*Nota 3: globals es un diccionario que cuenta con todas las variables globales que se utilizan en el programa actual.*

<a id="id2"> </a>
# Tipos de datos simples. Enteros, reales, booleanos y caracteres.

En programación, los datos son la información que se manipula. Los lenguajes de programación trabajan con diferentes tipos de datos para poder manipular la información. En este bloque nos enfocaremos en los tipos de datos simples en Python, que son los más básicos y fundamentales para cualquier programa.

Python es un lenguaje de programación orientado a objetos, lo que significa que cada valor en Python es un objeto. Un objeto es una estructura de datos que puede tener atributos (variables) y métodos (funciones). Los objetos en Python tienen un tipo que determina qué operaciones se pueden realizar con ellos.

## Tipos de datos simples

Los tipos de datos simples en Python se dividen en cuatro categorías: enteros, reales, booleanos y caracteres. En Python también existe un tipo de dato complejo, que nos permite trabajar con números complejos.

### Enteros (int)

Los enteros son números sin parte decimal. En Python, los enteros se representan con el tipo de datos int. Los enteros se pueden representar en notación decimal, binaria, octal y hexadecimal.

Por ejemplo, para definir una variable entera y mostrar su valor y tipo, podemos hacer lo siguiente:

In [ ]:
numero_entero = 10
print("Número entero:", numero_entero, "Tipo:", type(numero_entero))

### Operaciones Aritméticas con Enteros

Ahora te invitamos a realizar operaciones aritméticas básicas con números enteros en Python. Las operaciones aritméticas son fundamentales en cualquier lenguaje de programación y Python no es la excepción. Aquí vamos a ver ejemplos de suma, resta, multiplicación, división y exponenciación utilizando variables enteras.

Primero, definimos tres variables enteras `a`, `b`, y `c`, y una variable `res` que utilizaremos para almacenar los resultados de nuestras operaciones.

In [ ]:
a = 3
b = 5
c = 10
res = 0

Luego, realizamos las siguientes operaciones aritméticas:

- Suma: Calculamos la suma de a y b.
- Resta: Calculamos la resta de a menos b.
- Multiplicación: Calculamos la multiplicación de a y b.
- División: Calculamos la división de a entre b.
- Exponenciación: Calculamos a elevado a la potencia de b.

In [1]:
res=a+b
print(res)
res=a-b
print(res)
res=a*b
print(res)
res=a/b
print(res)
res=a**b
print(res)

NameError: name 'a' is not defined

Si ejecutaste el código, habrás notado que la salida no siempre es un número entero en todas las operaciones, especialmente en la división. Estos resultados que no son enteros se llaman números flotantes (los podemos identificar como decimales). Los números flotantes nos permiten realizar cálculos más precisos, especialmente en situaciones donde los resultados no son enteros.

A continuación, estudiaremos cómo trabajar con números flotantes en Python.

### Reales (float)

Los números reales, en el contexto de la programación, son aquellos que tienen una parte decimal. En Python, estos números se representan con el tipo de datos `float`.

Por ejemplo, para definir una variable real y mostrar su valor y tipo, podemos hacer lo siguiente:

In [2]:
numero_real = 3.1416
print("Número real:", numero_real, "Tipo:", type(numero_real))

Número real: 3.1416 Tipo: <class 'float'>


### Booleanos (bool)

Los booleanos son tipos de datos que solo pueden tomar dos valores: `True` (verdadero) o `False` (falso). En Python, los booleanos se representan con el tipo de datos `bool`.

Los valores booleanos se utilizan comúnmente en condiciones y pruebas de verdad. En Python, cualquier objeto puede evaluarse como verdadero o falso en un contexto booleano. Por ejemplo, los números no cero, las cadenas no vacías, las listas no vacías y otros objetos no vacíos se evalúan como `True`. Por otro lado, el número cero, las cadenas vacías, las listas vacías y otros objetos vacíos se evalúan como `False`.

Por ejemplo, para definir una variable booleana y mostrar su valor y tipo, podemos hacer lo siguiente:


In [3]:
verdadero = True
print("Booleano:", verdadero, "Tipo:", type(verdadero))

Booleano: True Tipo: <class 'bool'>


Los valores booleanos tambien pueden ser utilizados 

In [4]:
A = True
B = False
C = A and B 
print(C)

False


*Nota 5: Como operadores logicos solo existen 3, el and, or, y el not.
Faltando: el condicional, bicondicional*

### Caracteres (str)

En Python, los caracteres son representados como cadenas de longitud uno usando el tipo de datos `str`. Las cadenas de caracteres (`str`) son secuencias inmutables de caracteres utilizados para almacenar y manipular texto.

Para definir una variable de tipo carácter y mostrar su valor y tipo, podemos hacer lo siguiente:

In [5]:
caracter = 'a'
print("Carácter:", caracter, "Tipo:", type(caracter))

Carácter: a Tipo: <class 'str'>


Aunque Python no tiene un tipo de datos específico para caracteres, las cadenas de longitud uno pueden usarse para representar caracteres individuales. Además, las cadenas pueden contener varios caracteres, y se pueden manipular usando varios métodos y operaciones.

Por ejemplo, para evaluar el valor de verdad de una cadena, puedes hacer lo siguiente:

In [6]:
# Ejemplos de pruebas de valor de verdad con cadenas
print(bool("Hola"))  # True, porque la cadena no está vacía
print(bool(""))      # False, porque la cadena está vacía

True
False


### Complejos (complex)

Los números complejos son aquellos que tienen una parte real y una parte imaginaria. En Python, los números complejos se representan con el tipo de datos `complex`. Un número complejo se escribe como `a + bj`, donde `a` es la parte real y `b` es la parte imaginaria.

Por ejemplo, para definir una variable compleja y mostrar su valor y tipo, podemos hacer lo siguiente:

In [7]:
numero_complejo = 2+3j
print("Número complejo:", numero_complejo, "Tipo:", type(numero_complejo))

Número complejo: (2+3j) Tipo: <class 'complex'>


Python proporciona varias funciones y métodos para trabajar con números complejos. Por ejemplo, puedes obtener la parte real e imaginaria de un número complejo, así como su conjugado:

In [8]:
# Obtener la parte real e imaginaria
parte_real = numero_complejo.real
parte_imaginaria = numero_complejo.imag

# Obtener el conjugado
conjugado = numero_complejo.conjugate()

print("Parte real:", parte_real)
print("Parte imaginaria:", parte_imaginaria)
print("Conjugado:", conjugado)

Parte real: 2.0
Parte imaginaria: 3.0
Conjugado: (2-3j)


Los números complejos son útiles en muchas áreas de la matemática y la ingeniería, especialmente en el análisis de señales y la teoría de circuitos.

<a id="id3"> </a>
# Operadores Lógicos y de Comparación

Los operadores lógicos y de comparación son fundamentales en la programación, ya que permiten evaluar condiciones y realizar acciones basadas en el resultado de esas condiciones. En Python, tenemos operadores de comparación para comparar valores y operadores lógicos para combinar condiciones.

## Operadores de Comparación

Los operadores de comparación se utilizan para comparar dos valores y devuelven un valor booleano, es decir, `True` o `False`. Estos operadores nos permiten realizar comparaciones entre variables y valores, y son esenciales para el control del flujo de los programas mediante estructuras condicionales.

Los operadores de comparación en Python incluyen:

* **Igual (==)**: Verifica si dos valores son iguales.
* **Diferente (!=)**: Verifica si dos valores son diferentes.
* **Mayor que (>)**: Verifica si el valor de la izquierda es mayor que el valor de la derecha.
* **Menor que (<)**: Verifica si el valor de la izquierda es menor que el valor de la derecha.
* **Mayor o igual que (>=)**: Verifica si el valor de la izquierda es mayor o igual que el valor de la derecha.
* **Menor o igual que (<=)**: Verifica si el valor de la izquierda es menor o igual que el valor de la derecha.

Ejemplo de uso de operadores de comparación:

In [9]:
a = 5
b = 7

print("a == b:", a == b) # Falso, porque 5 no es igual a 7
print("a != b:", a != b) # Verdadero, porque 5 es diferente de 7
print("a > b:", a > b)   # Falso, porque 5 no es mayor que 7
print("a < b:", a < b)   # Verdadero, porque 5 es menor que 7
print("a >= b:", a >= b) # Falso, porque 5 no es mayor o igual que 7
print("a <= b:", a <= b) # Verdadero, porque 5 es menor o igual que 7

a == b: False
a != b: True
a > b: False
a < b: True
a >= b: False
a <= b: True


## Operadores Lógicos

Los operadores lógicos se utilizan para combinar condiciones y evaluar expresiones complejas. En Python, los operadores lógicos permiten tomar decisiones basadas en múltiples condiciones y son fundamentales para el control del flujo en la programación.

Los operadores lógicos en Python incluyen:

* **and**: Devuelve `True` si ambas condiciones son verdaderas (`True`), de lo contrario devuelve `False`.
* **or**: Devuelve `True` si al menos una de las condiciones es verdadera (`True`), de lo contrario devuelve `False`.
* **not**: Devuelve `True` si la condición es falsa (`False`), y `False` si la condición es verdadera (`True`).

### Ejemplo de uso de operadores lógicos

Supongamos que tenemos dos variables booleanas `x` y `y`. Podemos utilizar los operadores lógicos para combinar estas variables de diferentes maneras:


In [10]:
x = True
y = False

# Operador and: ambas condiciones deben ser verdaderas
print(x and y)  # False

# Operador or: al menos una de las condiciones debe ser verdadera
print(x or y)   # True

# Operador not: invierte el valor booleano
print(not x)    # False

False
True
False


Los operadores lógicos son especialmente útiles en estructuras de control como las sentencias if, donde se pueden evaluar múltiples condiciones para determinar el flujo del programa. Por ejemplo:

In [11]:
# Verificar si un número está en un rango específico
numero = 15
rango_minimo = 10
rango_maximo = 20

if numero >= rango_minimo and numero <= rango_maximo:
    print("El número está dentro del rango.")
else:
    print("El número está fuera del rango.")

El número está dentro del rango.


En este caso, el mensaje "El número está dentro del rango." se imprimirá si numero es mayor o igual a rango_minimo y menor o igual a rango_maximo.

Estos operadores permiten construir expresiones lógicas complejas y realizar evaluaciones detalladas, esenciales para la toma de decisiones en la programación.

<a id="id4"> </a>
# Comparaciones encadenadas

Una característica interesante de Python es la posibilidad de realizar comparaciones encadenadas, lo que permite escribir condiciones más compactas y legibles. Veamos un ejemplo de cómo podemos utilizar comparaciones encadenadas y luego plantearemos un desafío para poner en práctica este concepto.

En Python, se pueden encadenar comparaciones en una sola expresión. Por ejemplo, en lugar de escribir x < y and y < z, puedes escribir x < y < z. Esto es posible gracias a la evaluación de cortocircuito de Python, lo que significa que si la primera comparación devuelve False, el resto de las comparaciones no se evaluarán.

Ejemplo de comparaciones encadenadas:

In [13]:
a = 5
b = 10
c = 15

# Comprobar si a es menor que b y si b es menor que c
condicion_encadenada = a < b < c
print("Comparación encadenada (a < b < c):", condicion_encadenada) # Verdadero

Comparación encadenada (a < b < c): True


<a id="des1"> </a>
# Desafío 1: Clasificar numeros

Dado tres números distintos, clasificarlos en orden ascendente utilizando comparaciones encadenadas y operadores lógicos.

A continuación realizo el ejercicio.

In [17]:
a = 50
b = 100
c = 150

# Comprobar si a es menor que b y si b es menor que c
condicion_encadenada = a < b < c
print("Comparación encadenada (a < b < c):", condicion_encadenada) #verdadero

Comparación encadenada (a < b < c): True


<a id="des2"> </a>
# Desafío 2: Calcular promedio
Crear un programa que lea 5 calificaciones y calcule su promedio

Pasos para realizar el ejercicio:
Celda 1: crear una lista con las calificaciones. En vez de pedir al usuario que ingrese las notas, las escribimos nosotros en una lista. Así es más fácil probar el código.
Celda 2: Calcular el promedio.
- sum(calificaciones) suma todos los valores.
- len(calificaciones) cuenta cuántos elementos hay.
- Al dividir, obtenemos el promedio.
Celda 3: Mostrar el resultado en dos decimales.
- : .2f hace que el número se muestre con dos decimales.


In [19]:
# Paso 1: Crear una lista con calificaciones 
#En este caso , vamos a poner calificaciones directamente en la lista.
calificaciones = [7.5, 8.0, 6.5, 9.0, 7.0]
calificaciones 

#Paso 2: Calcular el promedio de las clificaciones
promedio = sum(calificaciones) / len(calificaciones)
promedio

#Paso 3: Mostrar el resultado formateado
print(f"El promedio de las calificaciones es: {promedio: .2f}")

El promedio de las calificaciones es:  7.60


Tipos de datos usados:
float → para representar calificaciones con decimales.
list → para guardar todas las calificaciones.
int → para recorrer posiciones con range(5).

Operadores usados:
+ → suma de los elementos.
/ → división para calcular el promedio.
sum() y len() → funciones para sumar y contar los elementos de la lista.

<a id="des3"> </a>
# Desafío 3: Convertidor de temperaturas
Realizar conversiones de una temperatura dada en Celsius a Fahrenheit y Kelvin utilizando operaciones directas sobre las variables. Los cálculos se realizarán directamente al asignar los valores a las variables.

Explicación del ejercicio: Celda 1: Elegimos una temperatura, por ejemplo 25ºC. Celda 2: Convertir a Fahrenheit. Celda 3: Converitr a Kelvin. Celda 4: Mostrar los resultados.
Fórmulas que vamos a utulizar: 
- Fahrenheit = (Celcius x 9/5) + 32
- Kelvin = Celcius + 273.15

In [32]:
# Definimos la temperatura en Celcius
celsius = 25
celsius

# Convertimos de Celcius a Fahrenheit
fahrenheit = (celsius * 9/5) + 32
fahrenheit

# Convertimos de Celsius a Kelvin
kelvin = celsius + 273.15
kelvin

# Mostramos todas las conversiones
print(f"Temperatura en Celsius: {celsius}ºc")
print(f"Equivalente en Fahrenheit: {fahrenheit: .2f}ºF")
print(f"Equivalente en Kelvin: {kelvin: .2f}K")

Temperatura en Celsius: 25ºc
Equivalente en Fahrenheit:  77.00ºF
Equivalente en Kelvin:  298.15K


Tipos de datos usados:
float → para representar temperaturas (ej. 25.0, 298.15).

Operadores usados:
* → multiplicación.
/ → división.
+ → suma.
f-string → para mostrar texto con variables dentro (f"...").

<a id="des4"> </a>
# Desafío 4: Verificar múltiplos de varios números
Objetivo: Determinar si un número dado es múltiplo de 2, 3, 5, 7, 9, 10, y 11 utilizando comparaciones y operadores lógicos.

*Descripción:*
Crea un programa que, dado un número, verifique si es múltiplo de los números mencionados. Para cada uno de estos números, el programa debe indicar si el número dado es o no un múltiplo de estos.

_Ejemplo de lo que se debe lograr:_
Supongamos que el número dado es 30. El programa deberá determinar:

¿Es 30 múltiplo de 2? Sí

¿Es 30 múltiplo de 3? Sí

¿Es 30 múltiplo de 5? Sí

¿Es 30 múltiplo de 7? No

¿Es 30 múltiplo de 9? No

¿Es 30 múltiplo de 10? Sí

¿Es 30 múltiplo de 11? No


Explicación paso a paso del ejercicio: Verificar si un número es múltimplo de otros, en mi ejemplo voy a utulizar el número 20.
Paso 1: Definir el número, esto simplemente guarda el numero en la variable (numero), que luego vamos a usar para hacer las comprobaciones. Paso 2: numero % 2 == 0, % es el operador módulo, que da el resto de una division, si el resto es 0, eso significa que el número es múltiplo. entonces 20%2 da 0: es múltiplo de 2, si hicieramos 20%3 que da 2, entonces no es múltiplo de 3.
Paso 3: 'Sí' if numero % 2 == 0 else 'No', esto es una forma compacta de escribir if; Si numero % 2 == 0 es verdadero devuelve 'Si', sino, devuelve 'No'.
Paso 4: Imprimir el resultado en una frase completa: print(f"¿Es {numero} múltiplo de 2? {'Sí if numero % 2 == 0 else 'No'}"), ¿Qué hace esto? usa una f-string (f"") para combinar texto con variables y cálculos. El resultado se vería así: ¿Es 20 múltiplo de 2? Sí. se repite el procedimiento con los otros números. 
Paso 5: Resultado final con numero = 20: un ejemplo sería ¿Es 20 múltiplo de 3? No, y asi sucesivamente hasta terminar la lista de números que tengamos.

In [40]:
# Números a verificar
2, 3 , 5, 7, 9, 10, 11

# Definimos el número que vamos a verificar
numero = 20

# Verificamos si es múltiplo de los distintos valores

print(f"¿Es {numero} múltiplo de 2? {'Sí' if numero % 2 == 0 else 'No'}")
print(f"¿Es {numero} múltiplo de 3? {'Sí' if numero % 3 == 0 else 'No'}")
print(f"¿Es {numero} múltiplo de 5? {'Sí' if numero % 5 == 0 else 'No'}")
print(f"¿Es {numero} múltiplo de 7? {'Sí' if numero % 7 == 0 else 'No'}")
print(f"¿Es {numero} múltiplo de 9? {'Sí' if numero % 9 == 0 else 'No'}")
print(f"¿Es {numero} múltiplo de 10? {'Sí'if numero % 10 == 0 else 'No'}")
print(f"¿Es {numero} múltiplo de 11? {'Sí'if numero % 11 == 0 else 'No'}")


¿Es 20 múltiplo de 2? Sí
¿Es 20 múltiplo de 3? No
¿Es 20 múltiplo de 5? Sí
¿Es 20 múltiplo de 7? No
¿Es 20 múltiplo de 9? No
¿Es 20 múltiplo de 10? Sí
¿Es 20 múltiplo de 11? No


¿Qué tipo de datos y operadores utilizaste?

Tipos de datos usados:
int → el número que se analiza (ej. 20).
str → para imprimir “Sí” o “No”

Operadores usados:
% (módulo) → para obtener el resto de una división.
== → comparación de igualdad.
if ... else (condicional inline).
f-string → para imprimir resultados con texto y variables.

<a id="des5"> </a>
# Desafío 5: Calculadora de Módulo

Escribe un programa que defina dos variables enteras a y b. Utiliza el operador de módulo (%) para calcular el residuo de la división de a entre b, y luego imprime el resultado. Asegúrate de que el programa maneje adecuadamente el caso en que b sea cero, imprimiendo un mensaje de error en ese caso.

<a id="ejemplos"></a>
# Problema

En Python, los identificadores son "sensibles a mayúsculas y minúsculas", lo que significa que numero y Numero son dos variables diferentes. Escribe un programa que defina dos variables, numero y Numero, les asigne valores diferentes y luego imprima ambos valores para demostrar esta característica.

In [ ]:
# Definir las variables con diferentes valores
numero = 10
Numero = 20

# Imprimir los valores de ambas variables
print("El valor de numero es:", numero)
print("El valor de Numero es:", Numero)
